File: scripts/fixture.test.ts
import { describe, expect, it } from "vitest";

import { formatLatency, getPercentile } from "./fixture.ts";

describe("formatLatency", () => {
  it("formats sub-millisecond values with 2 decimal places", () => {
    expect(formatLatency(0.45)).toBe("0.45ms");
    expect(formatLatency(0.01)).toBe("0.01ms");
    expect(formatLatency(0.999)).toBe("1.00ms");
  });

  it("formats 1-10ms values with 1 decimal place", () => {
    expect(formatLatency(1)).toBe("1.0ms");
    expect(formatLatency(2.34)).toBe("2.3ms");
    expect(formatLatency(9.99)).toBe("10.0ms");
  });

  it("formats 10-1000ms values as integers", () => {
    expect(formatLatency(10)).toBe("10ms");
    expect(formatLatency(45.6)).toBe("46ms");
    expect(formatLatency(999)).toBe("999ms");
  });

  it("formats values >= 1s in seconds", () => {
    expect(formatLatency(1000)).toBe("1.0s");
    expect(formatLatency(1500)).toBe("1.5s");
    expect(formatLatency(12345)).toBe("12.3s");
  });
});

describe("getPercentile", () => {
  it("returns 0 for empty array", () => {
    expect(getPercentile([], 0.5)).toBe(0);
  });

  it("returns the single element for a one-element array", () => {
    expect(getPercentile([42], 0.5)).toBe(42);
    expect(getPercentile([42], 0)).toBe(42);
    expect(getPercentile([42], 1)).toBe(42);
  });

  it("returns exact values at boundaries", () => {
    const sorted = [1, 2, 3, 4, 5];
    expect(getPercentile(sorted, 0)).toBe(1);
    expect(getPercentile(sorted, 1)).toBe(5);
  });

  it("returns the median for an odd-length array", () => {
    expect(getPercentile([1, 2, 3, 4, 5], 0.5)).toBe(3);
  });

  it("interpolates the median for an even-length array", () => {
    expect(getPercentile([1, 2, 3, 4], 0.5)).toBe(2.5);
  });

  it("interpolates fractional indices", () => {
    const sorted = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100];
    // p90 = index 9 * 0.9 = 8.1 → interpolate between sorted[8]=90 and sorted[9]=100
    const p90 = getPercentile(sorted, 0.9);
    expect(p90).toBeCloseTo(91, 0);
  });

  it("handles p99 on a large array", () => {
    const sorted = Array.from({ length: 1000 }, (_, i) => i + 1);
    const p99 = getPercentile(sorted, 0.99);
    // index = 0.99 * 999 = 989.01 → between sorted[989]=990 and sorted[990]=991
    expect(p99).toBeCloseTo(990.01, 1);
  });
});

File: cmd/dev/fixture_test.go
package main

import (
	"testing"

	"gotest.tools/v3/assert"
)

func TestFormatLatency(t *testing.T) {
	t.Parallel()

	cases := []struct {
		ms   float64
		want string
	}{
		// sub-millisecond, 2 decimal places
		{0.45, "0.45ms"},
		{0.01, "0.01ms"},
		{0.999, "1.00ms"},

		// 1-10ms, 1 decimal place
		{1, "1.0ms"},
		{2.34, "2.3ms"},
		{9.99, "10.0ms"},

		// 10-1000ms, integers
		{10, "10ms"},
		{45.6, "46ms"},
		{999, "999ms"},

		// >= 1s in seconds
		{1000, "1.0s"},
		{1500, "1.5s"},
		{12345, "12.3s"},
	}

	for _, tc := range cases {
		t.Run(tc.want, func(t *testing.T) {
			t.Parallel()
			assert.Equal(t, formatLatency(tc.ms), tc.want)
		})
	}
}

func TestGetPercentile(t *testing.T) {
	t.Parallel()

	t.Run("returns 0 for empty slice", func(t *testing.T) {
		t.Parallel()
		assert.Equal(t, getPercentile(nil, 0.5), 0.0)
	})

	t.Run("single element", func(t *testing.T) {
		t.Parallel()
		s := []float64{42}
		assert.Equal(t, getPercentile(s, 0), 42.0)
		assert.Equal(t, getPercentile(s, 0.5), 42.0)
		assert.Equal(t, getPercentile(s, 1), 42.0)
	})

	t.Run("exact boundaries", func(t *testing.T) {
		t.Parallel()
		s := []float64{1, 2, 3, 4, 5}
		assert.Equal(t, getPercentile(s, 0), 1.0)
		assert.Equal(t, getPercentile(s, 1), 5.0)
	})

	t.Run("median odd-length", func(t *testing.T) {
		t.Parallel()
		s := []float64{1, 2, 3, 4, 5}
		assert.Equal(t, getPercentile(s, 0.5), 3.0)
	})

	t.Run("median even-length interpolates", func(t *testing.T) {
		t.Parallel()
		s := []float64{1, 2, 3, 4}
		assert.Equal(t, getPercentile(s, 0.5), 2.5)
	})

	t.Run("p90 fractional index", func(t *testing.T) {
		t.Parallel()
		s := []float64{10, 20, 30, 40, 50, 60, 70, 80, 90, 100}
		// index = 9 * 0.9 = 8.1 -> interpolate between sorted[8]=90 and sorted[9]=100
		got := getPercentile(s, 0.9)
		assert.Assert(t, got > 90.9 && got < 91.1, "expected ~91, got %v", got)
	})

	t.Run("p99 on large array", func(t *testing.T) {
		t.Parallel()
		s := make([]float64, 1000)
		for i := range s {
			s[i] = float64(i + 1)
		}
		got := getPercentile(s, 0.99)
		assert.Assert(t, got > 990 && got < 991, "expected ~990.01, got %v", got)
	})
}
